User-tools: Add qualfication tool support for Onprem platform

nartal1 opened the pull request:
This fixes https://github.com/NVIDIA/spark-rapids-tools/issues/223
This PR adds qualification tool support in user-tools for Onprem platform.
Currently it supports only LOCAL mode and doesn't include cost savings.

Output format:
```
+----+---------------------------+-------------------------+----------------------+-----------------+-----------------+---------------+
|    | App Name                  | App ID                  | Speedup Based        |   Estimated GPU |   Estimated GPU |           App |
|    |                           |                         | Recommendation       |         Speedup |     Duration(s) |   Duration(s) |
|----+---------------------------+-------------------------+----------------------+-----------------+-----------------+---------------|
|  0 | TPC-DS Like Bench q24a    | app-20210507174503-2538 | Strongly Recommended |            4.08 |          238.23 |        970.98 |
|  1 | TPC-DS Like Bench q24b    | app-20210507180116-2539 | Strongly Recommended |            4.04 |          228.31 |        923.42 |
|  2 | TPC-DS Like Bench q14a    | app-20210319151533-1704 | Strongly Recommended |            3.50 |          216.17 |        756.04 |

```

<!--

Thank you for contributing to Tools of RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

amahussein commented on user_tools/docs/index.md:
```diff
@@ -132,6 +132,25 @@ The following table summarizes the commands supported for each cloud platform:
+|                  |               | serverless  |           unsupported                 |    N/A   |
+|                  |               +-------------+---------------------------------------+----------+
+|                  |               | cluster     |           unsupported                 |    N/A   |
++------------------+---------------+-------------+---------------------------------------+----------+
```
We may need to update the graph as well because the diagram does not show `onprem`.
Probably I should add the source of the overview.draw file to resources.

amahussein replied:
I created a PR adding the source file of the overview
https://github.com/NVIDIA/spark-rapids-tools/pull/304

amahussein commented on user_tools/docs/user-tools-onprem.md:
```diff
@@ -0,0 +1,130 @@
+# RAPIDS User Tools on OnPrem platform
+
+This is a guide for the RAPIDS tools for Apache Spark on OnPrem platform. At the end of this guide, the user will be able to run the RAPIDS tools to analyze the clusters and the applications running OnPrem.
```
I am not sure it is accurate to say `OnPrem platform`.
Maybe we should rephrase t o elaborate what we mean by OnPrem just in case someone can confuse it that it is some sort of a cloud product. for example
```suggestion
This is a guide for the RAPIDS tools for Apache Spark clusters that are provisioned manually (a.k.a onPrem). At the end of this guide, the user will be able to run the RAPIDS tools to analyze the clusters and the applications running OnPrem.
```

amahussein commented on user_tools/docs/user-tools-onprem.md:
```diff
@@ -0,0 +1,130 @@
+
+1. Store the Apache Spark event logs in local directory.
+2. User sets up the development machine:
+   1. configures Java
```
I know that this is the same line from other platforms. I just thought that the instruction is too vague.
We want to give some specific about what "configures java" mean. Or we replace it with something similar to  `on a machine with Java SDK 1.8+ installed1.

amahussein commented on user_tools/docs/user-tools-onprem.md:
````diff
@@ -0,0 +1,130 @@
+   ```
+   # define the wrapper cache directory if necessary
+   export RAPIDS_USER_TOOLS_CACHE_FOLDER=my_cache_folder
+   export EVENTLOGS=/LOGS_BUCKET/eventlogs/
````
`LOGS_BUCKET`  ?

nartal1 replied:
I hadn't specified LOGS_BUCKET but it was meant to be path_to_eventlogs. Updated it to reflect the same

amahussein commented on user_tools/docs/user-tools-onprem.md:
```diff
@@ -0,0 +1,130 @@
+   
+   spark_rapids_user_tools onprem qualification \
+      --eventlogs $EVENTLOGS \
+      --cpu_cluster $CLUSTER_NAME
```
We cannot use `CLUSTER_NAME` for on-prem. We need a cluster property file. right?

nartal1 replied:
You are correct. Initially, I thought it could be either of them but it will be a cluster property file for now. Updated it.

amahussein reviewed (commented):
I went quickly through the documentation.
I will resume my review later.

amahussein commented on user_tools/docs/user-tools-onprem.md:
```diff
@@ -0,0 +1,130 @@
+
+| Option                         | Description                                                                                                                                                                                                                                                                                               | Default                                                          ...
+|--------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------- ...
+| **cpu_cluster**                | The Onprem-cluster on which the Apache Spark applications were executed. Accepted values are an OnPrem-cluster name, or a valid path to the cluster properties file (json format)                                                                                                                         | N/A                                                              ...
```
should be a valid path only.

nartal1 replied:
Done.

mattahrens commented:
Can you file a follow-on issue for HDFS/S3 support for user tools on-prem along with expected scope needed?

mattahrens commented:
I tested this and got an error related to requiring cpu cluster: 

```
$ spark_rapids_user_tools onprem qualification --eventlog /tmp/spark-events
ERROR: The function received no value for the required argument: cpu_cluster
Usage: spark_rapids_user_tools onprem qualification CPU_CLUSTER <flags>
  optional flags:        --execution_cluster | --eventlogs | --local_folder |
                         --remote_folder | --gpu_cluster | --tools_jar |
                         --credentials_file | --filter_apps |
                         --gpu_cluster_recommendation | --jvm_heap_size |
                         --verbose
  additional flags are accepted
```

I don't think the user should have to require a CPU cluster for onprem, right?  They only need to supply the event log path.

amahussein commented:
> I don't think the user should have to require a CPU cluster for onprem, right? They only need to supply the event log path.

The qualification tool expects hardware description of the CPU-cluster
`cluster name` is not valid for on prem because there is no way API to map between cluster name and the actual cluster.
a cluster property file is the only way to get it to run.

@nartal1 what was the command used to generate the sample in PR's description?

mattahrens commented:
The java app for the qualification tool does not require any information about the CPU cluster, so we shouldn't require it for user tools onprem.

amahussein commented on user_tools/src/spark_rapids_pytools/rapids/qualification.py:
```diff
@@ -597,12 +607,36 @@ def __build_global_report_summary(self,
-        apps_pruned_df, prune_notes = self.__remap_columns_and_prune(all_apps)
-        recommended_apps = self.__get_recommended_apps(apps_pruned_df)
         apps_reshaped_df, per_row_flag = self.__apply_gpu_cluster_reshape(apps_pruned_df)
+        # OnPrem platform doesn't have pricing information. We do not calculate cost savings for
+        # OnPrem platform.
+        if pricing_config is None:
+            # Write to qualification_summary.csv file
+            if apps_reshaped_df is not None:
+                apps_reshaped_df = apps_reshaped_df.drop(columns=['Estimated Job Frequency (monthly)'])
+                self.logger.info('Generating GPU Estimated Speedup as %s', csv_out)
+                apps_reshaped_df.to_csv(csv_out, float_format='%.2f')
+
+            return QualificationSummary(comments=report_comments,
+                                        all_apps=apps_pruned_df,
+                                        recommended_apps=recommended_apps,
+                                        df_result=apps_reshaped_df,
+                                        irrelevant_speedups=speedups_irrelevant_flag,
+                                        sections_generators=[self.__generate_mc_types_conversion_report])
+
```
this part has a lot of repeated code that is going to be hard to maintain moving forward.
We may need to think what is the best way to reformat that block so we won't have multiple places when we are extending the code or debugging an issue.

nartal1 replied:
Thanks for the feedback. I have refactored the code to remove repeated code. PTAL.

amahussein commented on user_tools/src/spark_rapids_pytools/rapids/qualification.py:
```diff
@@ -58,6 +58,7 @@ class QualificationSummary:
     recommended_apps: pd.DataFrame = None
     df_result: pd.DataFrame = None
     irrelevant_speedups: bool = False
+    pricing_config: Any = None
```
is `pricing_config` is only used as a flag? or this is this defined as any to support another feature in the upcoming PR?

nartal1 replied:
It is used as a flag. If the pricing information is not available in config file, then we run qualification tool based on speedups and do not calculate/display cost savings.

amahussein commented:
> The java app for the qualification tool does not require any information about the CPU cluster, so we shouldn't require it for user tools onprem.

yes true, then #250 would require the argument. I believe this is why this PR did not get rid of it entirely.

nartal1 commented:
> > The java app for the qualification tool does not require any information about the CPU cluster, so we shouldn't require it for user tools onprem.
> 
> yes true, then #250 would require the argument. I believe this is why this PR did not get rid of it entirely.

That's correct. The reason I had kept CPU_CLUSTER arg was that it is a requirement for #250. Currently we don't read any info from the cluster property file. 
@mattahrens , @amahussein  -  Would it make sense to add CPU_CLUSTER arg mandatory only when `target_platform` (#250) is specified?

nartal1 commented:
@mattahrens, @amahussein - Updated the PR to run even when CPU_CLUSTER is not specified and filed follow on issue. Have addressed review comments. Could you PTAL.
